platform/security/cas/security-cas-client/src/main/java/org/codice/ddf/security/handler/cas/filter/FilteredRequest.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.handler.cas.filter;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import org.codice.ddf.configuration.SystemBaseUrl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A wrapped HttpServletRequest that is aware of the system root context, and returns the correct
+ * paths for both internal and external requests.
+ */
+public class FilteredRequest extends HttpServletRequestWrapper {
+  private static final Logger LOGGER = LoggerFactory.getLogger(FilteredRequest.class);
+
+  private final String contextPath;
+  private final String requestUri;
+  private final String requestUrl;
+
+  public FilteredRequest(HttpServletRequest request, String serverName) {
+    super(request);
+
+    // Wrapping FilteredRequests will result in repeated root contexts since the original request
+    // is already wrapped. Instead, just copy the wrapped request and systemBaseUrl references
+    if (request instanceof FilteredRequest) {
+      FilteredRequest filteredRequest = (FilteredRequest) request;
+      contextPath = filteredRequest.contextPath;
+      requestUri = filteredRequest.requestUri;
+      requestUrl = filteredRequest.requestUrl;
+    } else {
+      URL baseUrl = toBaseUrl(serverName);
+      contextPath =
+          isExternalRequest(baseUrl)
+              ? SystemBaseUrl.EXTERNAL.getRootContext().replaceFirst("/$", "")
```

Comment:
Stripping trailing slashes is so common we should have a library function for it.